plugins/out_es/es_conf.c

```diff
@@ -17,521 +17,899 @@
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "cannot configure path and/or pipeline");
+        return -1;
     }
 
-    /* Compress (gzip) */
-    tmp = flb_output_get_property("compress", ins);
-    ctx->compress_gzip = FLB_FALSE;
-    if (tmp) {
-        if (strcasecmp(tmp, "gzip") == 0) {
-            ctx->compress_gzip = FLB_TRUE;
-        }
+    ret = config_set_ra_id_key(ec->id_key, ec, ctx);
+    if (ret != 0) {
+        return -1;
     }
 
-    /* Prepare an upstream handler */
-    upstream = flb_upstream_create(config,
-                                   ins->host.name,
-                                   ins->host.port,
-                                   io_flags,
-                                   ins->tls);
-    if (!upstream) {
-        flb_plg_error(ctx->ins, "cannot create Upstream context");
-        flb_es_conf_destroy(ctx);
-        return NULL;
+    ret = config_set_es_action(ec->write_operation, ec->ra_id_key, ec->generate_id, ec,
+                               ctx);
+    if (ret != 0) {
+        return -1;
     }
-    ctx->u = upstream;
 
-    /* Set instance flags into upstream */
-    flb_output_upstream_set(ctx->u, ins);
+    ret = config_set_ra_prefix_key(ec->logstash_prefix_key, ec, ctx);
+    if (ret != 0) {
+        return -1;
+    }
 
-    /* Set manual Index and Type */
-    if (f_index) {
-        ctx->index = flb_strdup(f_index->value); /* FIXME */
+#ifdef FLB_HAVE_AWS
+    ret = flb_es_set_aws_unsigned_headers(ec);
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "cannot configure AWS unsigned headers");
+        return -1;
     }
 
-    if (f_type) {
-        ctx->type = flb_strdup(f_type->value); /* FIXME */
+    ret = flb_es_conf_set_aws_provider(
+            flb_output_get_property(FLB_ES_CONFIG_PROPERTY_AWS_EXTERNAL_ID, ctx->ins),
+            flb_output_get_property(FLB_ES_CONFIG_PROPERTY_AWS_ROLE_ARN, ctx->ins),
+            ec, ctx, config);
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "cannot configure AWS authentication");
+        return -1;
     }
+#endif
 
-    /* HTTP Payload (response) maximum buffer size (0 == unlimited) */
-    if (ctx->buffer_size == -1) {
-        ctx->buffer_size = 0;
+    return 0;
+}
+
+static int config_set_node_properties(struct flb_upstream_node *node,
+                                      struct flb_elasticsearch_config *ec,
+                                      struct flb_elasticsearch_config *base,
+                                      struct flb_elasticsearch *ctx,
+                                      struct flb_config *config)
+{
+    const char *tmp;
+    int ret;
+    const char *path;
+
+#ifdef FLB_HAVE_AWS
+    const char *aws_external_id = NULL;
+    const char *aws_role_arn = NULL;
+    int aws_provider_node = FLB_FALSE;
+#endif
+
+    /* Copy base configuration */
+    *ec = *base;
+    ec->own_index = FLB_FALSE;
+    ec->own_type = FLB_FALSE;
+    ec->own_cloud_user = FLB_FALSE;
+    ec->own_cloud_passwd = FLB_FALSE;
```

Comment:
We have to initialize with NULL instead of filling these variables with FLB_FALSE.

Reply:
Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

Reply:
> Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

We use C not C++. So, we can use initializing with NULL and free pattern. If you use C++, your comments is correct.
However, C does not have ownership on allocated memory. Thanks.